docs/data/data-grid/row-recipes/DetailPanelExpandCollapseAll.js

```diff
@@ -0,0 +1,134 @@
+import * as React from 'react';
+import Box from '@mui/material/Box';
+import IconButton from '@mui/material/IconButton';
+import UnfoldLessIcon from '@mui/icons-material/UnfoldLess';
+import UnfoldMoreIcon from '@mui/icons-material/UnfoldMore';
+import {
+  DataGridPro,
+  useGridApiContext,
+  useGridSelector,
+  gridDetailPanelExpandedRowIdsSelector,
+  GRID_DETAIL_PANEL_TOGGLE_COL_DEF,
+  gridDetailPanelExpandedRowsContentCacheSelector,
+} from '@mui/x-data-grid-pro';
+import {
+  randomCreatedDate,
+  randomCurrency,
+  randomEmail,
+  randomPrice,
+} from '@mui/x-data-grid-generator';
+
+export default function DetailPanelExpandCollapseAll() {
+  const getDetailPanelContent = React.useCallback(
+    ({ row }) => <Box sx={{ p: 2 }}>{`Order #${row.id}`}</Box>,
+    [],
+  );
+
+  const getDetailPanelHeight = React.useCallback(() => 50, []);
+
+  return (
+    <div style={{ height: 400, width: '100%' }}>
+      <DataGridPro
+        rows={rows}
+        columns={columns}
+        getDetailPanelContent={getDetailPanelContent}
+        getDetailPanelHeight={getDetailPanelHeight}
+      />
+    </div>
+  );
+}
+
+function CustomDetailPanelHeader() {
+  const apiRef = useGridApiContext();
+  const expandedRowIds = useGridSelector(
+    apiRef,
+    gridDetailPanelExpandedRowIdsSelector,
+  );
+  const rowsWithDetailPanels = useGridSelector(
+    apiRef,
+    gridDetailPanelExpandedRowsContentCacheSelector,
+  );
+
+  // little hack to always receive the keys in the correct type
+  const allRowIdsWithDetailPanels = Object.keys(rowsWithDetailPanels).map((key) =>
+    Number.isNaN(+key) ? key : +key,
+  );
+
+  const noDetailPanelsOpen = expandedRowIds.length === 0;
+
+  const expandOrCollapseAll = () => {
+    apiRef.current.setExpandedDetailPanels(
+      noDetailPanelsOpen ? allRowIdsWithDetailPanels : [],
```

Comment:
`allRowIdsWithDetailPanels` can be derived here, only when needed as opposed to always deriving it during render – just a small-ish optimization